### PR TITLE
Bugfixes for the HIP package on new versions of ROCm

### PR DESCRIPTION
**Pull request by bvanessen**
Fixed the rocm-dev-libs path when HIP is an external package for newer
versions of HIP.  Fixed rocm-path environment variable to point to the
ROCm path rather than the rocm-dev-libs variable.  Added a definition
of the HIP_CLANG_INCLUDE_PATH, which is not properly identified
because Spack wraps the compiler up in a Spack specific path, which
prevents proper detection of the Clang compiler headers.

### CODE REVIEW

**Comment by bvanessen** [comment]
@becker33 @scheibelp @haampie @balos1 Here are some fixes that enable the HIP package to work for my on LC.

**Review by adamjstewart — APPROVED** [review]

**Comment by adamjstewart** [comment]
@srekolam @arjun-raj-kuppala can you review?

**Comment by haampie** [comment]
Will take a look tomorrow!

File: var/spack/repos/builtin/packages/hip/package.py [hunk 1]
```diff
@@ -119,7 +124,9 @@ def set_variables(self, env):
         # it's necessary on runtime when using hiprtcCreateProgram and such
         env.set('LLVM_PATH', rocm_prefixes['llvm-amdgpu'])
         env.set('HIPCC_COMPILE_FLAGS_APPEND',
-                '--rocm-path={0}'.format(rocm_prefixes['device_lib_path']))
+                '--rocm-path={0}'.format(rocm_prefixes['rocm-path']))
+        env.set('HIP_CLANG_INCLUDE_PATH', '{0}/lib/clang/{1}/include'.format(
```
Comment (haampie) [inline var/spack/repos/builtin/packages/hip/package.py:128]: Why is this necessary? `hipcc` does the exact same thing:

https://github.com/ROCm-Developer-Tools/HIP/blob/2080cc113a2d767352b512b9d24c0620b6dee790/bin/hipcc#L203-L209

and cmake does not use the env variable. We already set HIP_CLANG_PATH to something that is not the compiler wrapper, so should be fine?

**Comment by srekolam** [comment]
i am trying out a few packages with this change ..right now i built rocblas@4.0.0,@3.8.0.. will update abt the progress in an hour or so.

**Comment by haampie** [comment]
@bvanessen can you share the output of the following without this change to `HIP_CLANG_INCLUDE_PATH`:

```
$ spack load hip
$ HIPCC_VERBOSE=2 hipcc --version
```
I'm curious why some of your variables are not set correctly with the external build

**Comment by srekolam** [comment]
this is breaking @3.7.0 for rocblas, rocrand packages.. 

20        Run Build Command(s):/usr/bin/gmake cmTC_60d2b/fast && /usr/bin/gmake  -f CMakeFiles/cmTC_60d2b.dir/build.make CMakeFi
           les/cmTC_60d2b.dir/build
     21        gmake[1]: Entering directory '/tmp/root/spack-stage/spack-stage-rocblas-3.7.0-ubamgh57ba7ufnt4amy5xvez3k5b7sni/spack-b
           uild-ubamgh5/CMakeFiles/CMakeTmp'
     22        Building CXX object CMakeFiles/cmTC_60d2b.dir/testCXXCompiler.cxx.o
     23        /spack/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/hip-3.7.0-itxcp66ohqd44ocz3wj7l5qjwcvvesjm/bin/hipcc    -o CMa
           keFiles/cmTC_60d2b.dir/testCXXCompiler.cxx.o -c /tmp/root/spack-stage/spack-stage-rocblas-3.7.0-ubamgh57ba7ufnt4amy5xvez3k
           5b7sni/spack-build-ubamgh5/CMakeFiles/CMakeTmp/testCXXCompiler.cxx
  >> 24    clang-11: error: unsupported option '--rocm-path=/spack/opt/spack/linux-centos8-skylake_avx512/gcc-8.3.1/hip-3.7.0-itxcp66
           ohqd44ocz3wj7l5qjwcvvesjm'
  >> 25        gmake[1]: *** [CMakeFiles/cmTC_60d2b.dir/build.make:85: CMakeFiles/cmTC_60d2b.dir/testCXXCompiler.cxx.o] Error 1
     26        gmake[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-rocblas-3.7.0-ubamgh57ba7ufnt4amy5xvez3k5b7sni/spack-bu
           ild-ubamgh5/CMakeFiles/CMakeTmp'
     27        gmake: *** [Makefile:140: cmTC_60d2b/fast] Error 2

**Comment by haampie** [comment]
@srekolam that must have been the case already since `--rocm-path` was around since before this pr.

You might want to review #21852 instead which fixes your issue and includes this PR except for the `HIP_CLANG_INCLUDE_PATH` variable.

**Comment by bvanessen** [comment]
> @bvanessen can you share the output of the following without this change to `HIP_CLANG_INCLUDE_PATH`:
> 
> ```
> $ spack load hip
> $ HIPCC_VERBOSE=2 hipcc --version
> ```
> 
> I'm curious why some of your variables are not set correctly with the external build

Here is what I get 
```
lbann-test-hip-zen2] (2) vanessen@corona211 [lbann_bvanessen.git] %> HIPCC_VERBOSE=2 hipcc --version
HIP_PATH=/opt/rocm-4.0.0
HIP_PLATFORM=hcc
HIP_COMPILER=clang
HIP_RUNTIME=ROCclr
ROCM_PATH=/opt/rocm-4.0.0
HIP_ROCCLR_HOME=/opt/rocm-4.0.0
HIP_CLANG_PATH=/opt/rocm-4.0.0/llvm/bin
HIP_CLANG_INCLUDE_PATH=/opt/rocm-4.0.0/llvm/lib/clang/12.0.0/include
HIP_INCLUDE_PATH=/opt/rocm-4.0.0/include
HIP_LIB_PATH=/opt/rocm-4.0.0/lib
DEVICE_LIB_PATH=/opt/rocm-4.0.0/amdgcn/bitcode
HIP version: 4.0.20496-4f163c68
clang version 12.0.0 (/data/jenkins_workspace/centos_pipeline_job_4.0/rocm-rel-4.0/rocm-4.0-23-20201214/7.7/external/llvm-project/clang dac2bfceaa8d4a90257dc8a6d58f268e172ce00e)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/rocm-4.0.0/llvm/bin
```
However, I also see the following build errors:
```
  >> 395    CMake Error in tests/CMakeLists.txt:
     396      Target "lbann" contains relative path in its INTERFACE_INCLUDE_DIRECTORIES:
     397    
     398        "HIP_CLANG_INCLUDE_PATH-NOTFOUND/.."
```
So I am not sure what is happening, but explicitly setting the environment variable allowed it to work for me.

**Comment by bvanessen** [comment]
> @srekolam that must have been the case already since `--rocm-path` was around since before this pr.
> 
> You might want to review #21852 instead which fixes your issue and includes this PR except for the `HIP_CLANG_INCLUDE_PATH` variable.

I was just able to test the current PR 21822 without manually setting the environment variable and it doesn't work for me.  I can try the new PR 21852 later, but I don't expect it to work without that variable.

**Comment by haampie** [comment]
Ah, this issue is caused by find_package(hip) which is bad at locating clang.

Can you try the following: https://github.com/spack/spack/blob/6aacdbbc76d5859a14baa991b05051b257f7dc1e/var/spack/repos/builtin/packages/spfft/package.py#L78

**Comment by haampie** [comment]
The hip cmake script tries to link to clang's builtins lib even for host code (the hip::host target).

In spack there is a patch to drop that behavior, since it was intended to only add half precision support to device code (the hip::device target). Obviously if you use an external hip install you don't have that patch.

There's an open PR to drop this in hip itself, so you can show your support: https://github.com/ROCm-Developer-Tools/HIP/pull/2219

**Comment by bvanessen** [comment]
> Ah, this issue is caused by find_package(hip) which is bad at locating clang.
> 
> Can you try the following:
> 
> https://github.com/spack/spack/blob/6aacdbbc76d5859a14baa991b05051b257f7dc1e/var/spack/repos/builtin/packages/spfft/package.py#L78

Yes if I specifically set the HIP CXX compiler it works without defining the CLANG include path in the HIP package.

**Comment by haampie** [comment]
Closing this in favor of #21852
